pkg/apis/configuration/v1alpha1/kong_custom_entity_types.go

```diff
@@ -0,0 +1,94 @@
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	KongCustomEntityKind = "KongCustomEntity"
+)
+
+type KongEntityScope string
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced,shortName=kce,categories=kong-ingress-controller,path=kongcustomentities,singular=kongcustomentity
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Entity Type",type=string,JSONPath=`.type`,description="type of the Kong entity"
+// +kubebuilder:validation:XValidation:rule="self.scope == oldSelf.scope",message="The scope field is immutable"
+// +kubebuilder:validation:XValidation:rule="self.type == oldSelf.type",message="The type field is immutable"
+// +kubebuilder:validation:XValidation:rule="!(self.type in ['services','routes','upstreams','targets','plugins','consumers','consumer_groups'])",message="The type field cannot be known Kong entity types"
+// +kubebuilder:validation:XValidation:rule="!(self.scope == 'independent' && has(self.parentRef)) && !(self.scope == 'attached' && !has(self.parentRef))",message="attached KongCustomEntity must have parentRef; independent KongCustomEntity must not have parentRef"
+// REVIEW: put all fields other than "status" under the "spec"?
+
+// KongCustomEntity defines a "custom" Kong entity that KIC cannot support the entity type directly.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Enum=independent;attached
+	// REVIEW: We can get the schema of entities from Kong gateway admin APIs including fields requiring foreign reference.
+	// Is the "Scope" field really required?
+
+	// Scope means whether the entity can be specified independently.
+	// "independent" means that the entity does not depend on other entities;
+	// "attached" means that the entity has "foreign" fields referring to other entities.
+	Scope KongEntityScope `json:"scope"`
```

Comment:
I think we can probably remove this field. The `Scope` field is intended to be used for telling KIC to find for "attached" entities (fields that requires a foreign key), for translating into configuration and filling mock data for validation. Since we can get the schema from Kong gateway's admin API, this field is not so required.

Reply:
Removed the `Scope` field.